### PR TITLE
Autocode java renderers

### PR DESCRIPTION
**Pull request by premprakash90**
@theRobinator please review and let me know if anything needs to be changed.

### CODE REVIEW

File: language_utils/java.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+"""
+Closure utilities.
+"""
+
+PRIMITIVE_TYPES = ['byte', 'short', 'int', 'long', 'float', 'double', 'boolean', 'char', 'String']
+BUILT_IN_VARS = {'Infinity', 'NaN', 'null', 'undefined', 'true', 'false', 'window', 'document', 'this',  # Values
```
Comment (theRobinator) [inline language_utils/java.py]: These are not the Java builtins.

File: language_utils/java.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    return name.endswith('_')
+
+
+def sort_fields(fields):
+    """ Sort an iterable of fields into an ordered list. Sort order is static
+        then instance, public then private, and inheritDoc, typed, and non-typed
+    """
+    def sortKey(item):
+        return (not item.const, not item.static, item.type is None, item.has_prop('private'), item.name)
```
Comment (theRobinator) [inline language_utils/java.py:68]: item.type can never be None in Java.

File: renderers/java/classrenderer.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        inheritance_string += ' implements %s' % ', '.join(cls.implements)
+    result.append("class %s%s {" % (cls.name, inheritance_string))
+    indent_result = []
+
+    # Fields
+    if len(cls.fields) > 0:
+        indent_result.extend(x.render(cls) for x in utils.sort_fields(cls.fields.values()))
+        indent_result.append("\n")
+
+    from autocode.primitives import Function
```
Comment (theRobinator) [inline renderers/java/classrenderer.py:38]: You should check for a constructor function to be set here. Java has default constructors.

File: renderers/java/enumrenderer.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+def render(enum, owner):
+    if enum.value_type is None:
+        raise Exception('Enums must have type')
+
+    enum.add_prop('enum', enum.value_type.name)
```
Comment (theRobinator) [inline renderers/java/enumrenderer.py]: I don't think this is a legitimate javadoc tag. This means that we can remove the check for enum.value_type above too.

File: renderers/java/fieldrenderer.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import re
+
+
+def render(field, owner):
+    if field.has_prop('var'):
```
Comment (theRobinator) [inline renderers/java/fieldrenderer.py]: @var isn't a javadoc tag.

File: renderers/java/fieldrenderer.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        field.remove_all_props('var')
+    if field.type is None:
+        raise Exception('Could not determine the type of the field %s' % field.type)
+
+    field.add_prop('var', field.type.name)
```
Comment (theRobinator) [inline renderers/java/fieldrenderer.py]: Same thing here.

File: renderers/java/fieldrenderer.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    if field.const:
+        field.remove_all_props('access')
+
+    if field.description is None and len(field.props) == 0:
```
Comment (theRobinator) [inline renderers/java/fieldrenderer.py]: Looks like this if statement is backwards. It's only rendering the comment if there's nothing to render.

File: renderers/java/fieldrenderer.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        result = ['/**', field.render_comment(), ' */']
+    else:
+        result = []
+
+    if field.value is not None:
+        value_str = ' = %s' % field.value
+    else:
+        value_str = ''
+
+    if field.const:
```
Comment (theRobinator) [inline renderers/java/fieldrenderer.py:19]: The final and static versions are missing visibility below.

File: renderers/java/functionrenderer.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+def render(func, owner):
+    """ Return a string containing what this would look like in code. """
+    param_string_parts = []
+    for param in func.params:
+
+        if param.type is None:
+            raise Exception('Could not determine the type of the param %s' % param.name)
+
+        if func.return_type is None and func.name != owner.name:
+            raise Exception('Java function must have return type except default constructor')
```
Comment (theRobinator) [inline renderers/java/functionrenderer.py]: Please change this error to say "All Java functions other than constructors must have return types"

File: renderers/java/valuerenderer.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        return 'new %s[]{%s};' % (java_type, ', '.join(result))
+    elif type(value) == dict:
+        result = []
+        if len(value) <= 0:
+            raise Exception('Empty python dict. Cannot convert to java dict')
+
+        key, val = value.popitem()
+        key_type = type(key)
+        value_type = type(val)
+        result.append('     put(%s,%s);' % (key, render(val)))
```
Comment (theRobinator) [inline renderers/java/valuerenderer.py]: In this and on line 40 you need to render the key as well.

**Comment by theRobinator** [comment]
It looks like you're also missing an enum for java in languages.py.

**Comment by premprakash90** [comment]
@theRobinator please review the fixes. I am not sure how to identify the built in vars in java. I added the built in classes and exception.

**Comment by theRobinator** [comment]
The latest version of this crashes when generating v5/test/object/get. Also, the PHP layer crashes in the propertyrenderer because EMPTY_COMMENT is not defined.

Run this to generate all the endpoints in java (change the end to "full" or "php" to see the other crash):
./bloom list api | head -21 | tail -20 | sed 's/^ *//' | xargs -n 1 ./bloom java

**Pull request merged**